pydantic/schema.py

```diff
@@ -971,15 +972,20 @@ def multitypes_literal_field_for_schema(values: Tuple[Any, ...], field: ModelFie
 
 
 def encode_default(dft: Any) -> Any:
-    if isinstance(dft, Enum):
+    from .main import BaseModel
+
+    if isinstance(dft, BaseModel) or is_dataclass(dft):
+        dft = cast('dict[str, Any]', pydantic_encoder(dft))
+    elif isinstance(dft, Enum):
         return dft.value
     elif isinstance(dft, (int, float, str)):
         return dft
     elif isinstance(dft, (list, tuple)):
         t = dft.__class__
         seq_args = (encode_default(v) for v in dft)
         return t(*seq_args) if is_namedtuple(t) else t(seq_args)
-    elif isinstance(dft, dict):
+
+    if isinstance(dft, dict):
```

Comment:
Do we need this change here?

Reply:
Yes. Notice the change above: we transform models and dataclasses to `dict` in first `if` and need this to be a separate one to work.

Reply:
Any ideas how to make this more explicit and obvious, btw?

Reply:
maybe just a comment with the above explanation.

Reply:
I would add the first if as a separate if and move the `isinstance(dft, dict)` check to the top.

Here is my suggestion(I haven't tested it locally):
```python
def encode_default(dft: Any) -> Any:
    from .main import BaseModel

    if isinstance(dft, BaseModel) or is_dataclass(dft):
        dft = cast('dict[str, Any]', pydantic_encoder(dft))

    if isinstance(dft, dict):
        return {encode_default(k): encode_default(v) for k, v in dft.items()}
    elif isinstance(dft, Enum):
        return dft.value
    elif isinstance(dft, (int, float, str)):
        return dft
    elif isinstance(dft, (list, tuple)):
        t = dft.__class__
        seq_args = (encode_default(v) for v in dft)
        return t(*seq_args) if is_namedtuple(t) else t(seq_args)
    elif dft is None:
        return None
    else:
        return pydantic_encoder(dft)

```

Reply:
Happy to accept suggestions on phrasing as well :)
```suggestion
    # we can end up here either from first condition or if none of above conditions were met
    if isinstance(dft, dict):
```

Reply:
Thanks for suggestion, @hramezani, I like your approach! I think I was hesitant on reordering at first, but it seems like the best move here, indeed.